### ISSUE TITLE
[TypeScript] Dynamically-compiled inputs have incorrect imported file types

### ISSUE THREAD

**Issue by Paril** [rollup/plugins#909]
- Rollup Plugin Name: TypeScript
- Rollup Plugin Version: Version
8.2.1
- Rollup Version:2.51.2
- Operating System (or Browser): W10
- Node Version: v14.8.0
- Link to reproduction: https://replit.com/@Paril/rollup-repro

### Expected Behavior

Rollup + TypeScript should be able to properly handle TS imports from inputs that are "synthetic", so to speak.

### Actual Behavior

With a "synthetic" input, it seems to assume all of the files are JS, even if the file is definitely absolutely a `.ts` file.

### Additional Information

Okay, so this one is a bit hard to explain because we have a completely in-house "prep" system for compiling our JS from dev to production.

Long story short, we have a prep system that can do various things, like merging/modifying JS files before passing them to rollup. Rollup is given a proper directory to work with (for imports, etc) but the input is entirely synthetic (that is to say, it's not a file on disk, it's constructed beforehand by prep, and then we use an in-house Rollup plugin to resolve the input to the actual contents of the JS file (which is always JS and always synethetic/in memory only, not on disk).

I started to want to work a bit with TypeScript in my codebase. I figured/hoped that all that the TypeScript plugin would do is just be a proper handler for .ts imports and that'd be it, but there seems to be some other things going on here that seem extremely overkill for our purposes. I went through a few iterations of various configuration options, and finally when I landed on something that I thought would be "perfect", it seems to just ignore the fact that the imported file is TS and assumes it's JS, despite it having a ts extension.

There's a second piece to this puzzle that I *think* is related to the fact that the plugin seems to launch a whole `watch` process or something from TSC. I don't need this at all; I'm not keeping a process alive or anything, this is a one-shot transpile. Locally in our repo, if I remove the `allowJs: false, checkJs: false` lines, the code does transpile properly, but I get thrown a lot of warnings about files that aren't even imported in the Rollup input file(s):

```
[6:20:43 am] [warning] @rollup/plugin-typescript TS5055: Cannot write file 'G:/xampp2/htdocs/PMC/core/java/dynamic_header/BitList.js' because it would overwrite input file.
[6:20:43 am] [warning] @rollup/plugin-typescript TS5055: Cannot write file 'G:/xampp2/htdocs/PMC/core/java/dynamic_header/Color.js' because it would overwrite input file.
[6:20:43 am] [warning] @rollup/plugin-typescript TS5055: Cannot write file 'G:/xampp2/htdocs/PMC/core/java/dynamic_header/Deque.js' because it would overwrite input file.
```

Ad infinitum, for seemingly every single .js file that exists in our repo. I can't reproduce this with the REPL at all, so I'm not sure what's causing this one, but it seems like TSC is wanting to compile all of our files despite `noEmit` being true. No idea what this one is all about. The process takes forever because of this issue, but the unfortunate part is it actually *works*, unlike if I turn checkJs/allowJs off, where it then just thinks every .ts import is .js and throws errors. It also seems to launch nearly 2x more processes than it should *if* I have the TS plugin enabled at all, which completely eats up my CPU, whereas without the TS plugin it doesn't do this. Can I disable this "watch" functionality somehow and just have simple string->string transpilation?

In the repl, you can see that I have the TS importing plugin enabled; if you change `input` directly to `main.js`, it works fine. If you change the return value of `load(id)` to import from `./src/test.ts`, it seems to work, but this shouldn't be necessary (and will throw errors in VSCode/tsc anyways; `An import path cannot end with a '.ts' extension. Consider importing '../../lib/Blobs.js' instead. ts(2691)`. Rollup seems to allow it, but, I can't use this workaround since it breaks our IntelliSense) because the file itself is already .ts and Rollup should be able to pick up on that (it does if main.js is used, but not if it's a synthetic input file).

**Comment by shellscape** [comment]
This may be a case where your usage is edge, and the plugin can't support it, in which case a third party TS rollup plugin might be better for you. Fixing this is probably going to require direct, intimate knowledge of your case, which means this probably won't get picked up by maintainer and will need community contribution.

**Comment by Paril** [comment]
> This may be a case where your usage is edge, and the plugin can't support it, in which case a third party TS rollup plugin might be better for you. Fixing this is probably going to require direct, intimate knowledge of your case, which means this probably won't get picked up by maintainer and will need community contribution.

Thanks for the reply.

So, I tested this with an extremely simple transpile custom plugin, and that works great. I think the issue is just down to how this plugin handles TS. I don't know really why it's using watch even when roll-up isn't in watch mode, it's like it runs tsc on the entire project when all I want it to do is handle ts files when they're imported. Using a simple ts.transpile wrap also kept my compile times as they were before (they were about 22s total with this plugin, but only the usual 8s if I transpiled on demand). Maybe I misread the plugin here and thought it was just a simple transpiler and not a whole watcher or something.
